File: Behavioral/State/DIY.java
package State;


public class DIY {
    public static void main(String[] args) {
        var pc = new PlayerController();
        pc.setCurrentTool(new Walking());
        pc.keyUp();
        pc.keyDown();
        pc.keyLeft();
        pc.keyRight(); 
    }
}


class PlayerController {
    private Keys currentTool;

    public void keyUp() {
        currentTool.keyUp();
    }

    public void keyDown() {
        currentTool.keyDown();
    }

    public void keyLeft() {
        currentTool.keyLeft();
    }

    public void keyRight() {
        currentTool.keyRight();
    }

    public Keys getCurrentTool() {
        return currentTool;
    }

    public void setCurrentTool(Keys currentTool) {
        this.currentTool = currentTool;
    }
}

interface Keys {
    void keyUp();
    void keyDown();
    void keyLeft();
    void keyRight();
}

class Walking implements Keys {
    @Override
    public void keyUp() {
        System.out.println("Player is jumping.");
    }

    @Override
    public void keyDown() {
        System.out.println("Player is rolling.");
    }

    @Override
    public void keyLeft() {
        System.out.println("Player is walking left.");
    }

    @Override
    public void keyRight() {
        System.out.println("Player is walking right.");
    }
}

class Inventory implements Keys {
    @Override
    public void keyUp() {
        System.out.println("Move highlighted box up");
    }

    @Override
    public void keyDown() {
        System.out.println("Move highlighted box down");
    }

    @Override
    public void keyLeft() {
        System.out.println("Move highlighted box left");
    }

    @Override
    public void keyRight() {
        System.out.println("Move highlighted box right");
    }
}

File: TS/Behavioral/State/DIY.ts
interface Keys {
  keyUp(): void;
  keyDown(): void;
  keyLeft(): void;
  keyRight(): void;
}

class Walking implements Keys {
  keyUp(): void {
    console.log("Player is jumping.");
  }

  keyDown(): void {
    console.log("Player is rolling.");
  }

  keyLeft(): void {
    console.log("Player is walking left.");
  }

  keyRight(): void {
    console.log("Player is walking right.");
  }
}

class Inventory implements Keys {
  keyUp(): void {
    console.log("Move highlighted box up");
  }

  keyDown(): void {
    console.log("Move highlighted box down");
  }

  keyLeft(): void {
    console.log("Move highlighted box left");
  }

  keyRight(): void {
    console.log("Move highlighted box right");
  }
}

class PlayerController {
  private currentTool: Keys;

  keyUp(): void {
    this.currentTool.keyUp();
  }

  keyDown(): void {
    this.currentTool.keyDown();
  }

  keyLeft(): void {
    this.currentTool.keyLeft();
  }

  keyRight(): void {
    this.currentTool.keyRight();
  }

  getCurrentTool(): Keys {
    return this.currentTool;
  }

  setCurrentTool(currentTool: Keys): void {
    this.currentTool = currentTool;
  }
}

// The Client
const pc = new PlayerController();
pc.setCurrentTool(new Walking());
pc.keyUp();
pc.keyDown();
pc.keyLeft();
pc.keyRight();
